File: src/NnCase.IR/Operators/Transpose.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace NnCase.IR.Operators
{
    public class Transpose : Node
    {
        public InputConnector Input { get; }

        public OutputConnector Output { get; }

        public Shape Perm { get; }

        public Transpose(DataType type, Shape inputShape, Shape perm)
        {
            Perm = perm;
            Input = AddInput("input", type, inputShape);
            Output = AddOutput("output", type, ShapeUtility.GetTransposedShape(inputShape, perm));
        }
    }
}

File: src/ir/ops/transpose.cpp
/* Copyright 2019 Canaan Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <ir/ops/transpose.h>
#include <ir/op_utils.h>
#include <xtensor/xarray.hpp>

using namespace nncase;
using namespace nncase::ir;

transpose::transpose(datatype_t type, shape_t input_shape, axis_t perm)
    : perm_(std::move(perm))
{
    add_input("input", type, input_shape);
    add_output("output", type, get_transposed_shape(input_shape, perm));
}
